classes/openxt-image-disk.bbclass

```diff
@@ -24,87 +42,91 @@ CONVERSION_DEPENDS_disk = "syslinux \
-	echo "LABEL boot" >> ${BOOTDIR}/${CFGNAME}
-	echo "KERNEL /${KERNEL_IMAGETYPE}" >> ${BOOTDIR}/${CFGNAME}
-	echo "APPEND LABEL=boot ${KERNEL_CMDLINE}" >> ${BOOTDIR}/${CFGNAME}
-	if [ -n "${INITRD}" ]; then
-		echo "INITRD /initrd" >> ${BOOTDIR}/${CFGNAME}
-	fi
+SYSLINUX_CFG ?= "${S}/syslinux.cfg"
+GRUB_CFG ?= "${S}/grub.cfg"
 
-	BOOTKB=$( du -bks $BOOTDIR | cut -f 1 )
-	BOOTKB=$( expr $BOOTKB + ${BOOTDD_EXTRA_SPACE} )
+build_boot_dd() {
+	HDDDIR="${S}/hdd/boot"
+	HDDIMG="${S}/hdd.image"
+	ROOTFS_TYPE="${1}"
+	ROOTFS="${IMAGE_NAME}${IMAGE_NAME_SUFFIX}.${ROOTFS_TYPE}"
+	IMAGE="${ROOTFS}.disk"
 
-	BOOTDD_VOLUME_ID="BOOT"
-	# Remove it since mkdosfs would fail when it exists
-	rm -f $BOOTIMG
-	mkdosfs -n ${BOOTDD_VOLUME_ID} -S 512 -C $BOOTIMG $BOOTKB
-	mcopy -i $BOOTIMG -s $BOOTDIR/* ::/
+	populate_kernel $HDDDIR
+	populate_initrd $HDDDIR
 
-	# syslinux installs ldlinux.sys & ldlinux.c32
-	syslinux $BOOTIMG
-	chmod 644 $BOOTIMG
+	if [ "${PCBIOS}" = "1" ]; then
+		syslinux_hddimg_populate $HDDDIR
+	fi
+	if [ "${EFI}" = "1" ]; then
+		efi_hddimg_populate $HDDDIR
+	fi
 
-	ROOTKB=$( du -Lbks $ROOTIMG | cut -f 1 )
-	TOTALSIZE=$( expr $BOOTKB + $ROOTKB )
-	END1=$( expr $BOOTKB \* 1024 )
-	START2=$( expr $END1 + 512 )
-	END2=$( expr \( $ROOTKB \* 1024 \) + $END1 )
+	BLOCKS=`du -bks $HDDDIR | cut -f 1`
+	BLOCKS=`expr $BLOCKS + ${BOOTDD_EXTRA_SPACE}`
 
-	echo $ROOTKB $TOTALSIZE $END1 $START2 $END2
+	# Remove it since mkdosfs would fail when it exists
+	rm -f $HDDIMG
+	mkdosfs -n ${BOOTDD_VOLUME_ID} -S 512 -C $HDDIMG $BLOCKS 
+	mcopy -i $HDDIMG -s $HDDDIR/* ::/
+
+	if [ "${PCBIOS}" = "1" ]; then
+		syslinux_hdddirect_install $HDDIMG
+	fi	
+	chmod 644 $HDDIMG
+
+	ROOTFSBLOCKS=`du -Lbks ${ROOTFS} | cut -f 1`
+	TOTALSIZE=`expr $BLOCKS + $ROOTFSBLOCKS`
+	END1=`expr $BLOCKS \* 1024`
+	END2=`expr $END1 + 512`
+	END3=`expr \( $ROOTFSBLOCKS \* 1024 \) + $END1`
+
+	echo $ROOTFSBLOCKS $TOTALSIZE $END1 $END2 $END3
 	rm -rf $IMAGE
 	dd if=/dev/zero of=$IMAGE bs=1024 seek=$TOTALSIZE count=1
 
 	parted $IMAGE mklabel msdos
 	parted $IMAGE mkpart primary fat16 0 ${END1}B
-	parted $IMAGE unit B mkpart primary ext2 ${START2}B ${END2}B
-	parted $IMAGE set 1 boot on
+	parted $IMAGE unit B mkpart primary ext2 ${END2}B ${END3}B
+	parted $IMAGE set 1 boot on 
 
 	parted $IMAGE print
 
-	# write a disk signature
-	echo -n "disk" | dd of=$IMAGE bs=1 seek=440 conv=notrunc
+	awk "BEGIN { printf \"$(echo ${DISK_SIGNATURE} | sed 's/\(..\)\(..\)\(..\)\(..\)/\\x\4\\x\3\\x\2\\x\1/')\" }" | \
+		dd of=$IMAGE bs=1 seek=440 conv=notrunc
+
+	OFFSET=`expr $END2 / 512`
+	if [ "${PCBIOS}" = "1" ]; then
+		dd if=${STAGING_DATADIR}/syslinux/mbr.bin of=$IMAGE conv=notrunc
+	fi
+
+	dd if=$HDDIMG of=$IMAGE conv=notrunc seek=1 bs=512
+	dd if=${ROOTFS} of=$IMAGE conv=notrunc seek=$OFFSET bs=512
+} 
 
-	OFFSET=$( expr $START2 / 512 )
+validate_disk_signature() {
+	if [ $(/bin/echo -n ${DISK_SIGNATURE}|wc -c) -le 8 ]; then
+		if [ $(expr "x${DISK_SIGNATURE}" : "x[A-Za-z0-9]*$") -gt 0 ]; then
```

Comment:
Upstream but if we want to rewrite in what we think is a more optimized approach, I am good with that.

Reply:
Optimized or not, this test is insufficient. Inappropriate values can get passed it.
Might worth being submitted upstream as well then.

Reply:
True, because even their own docs state that it should be a hexdigit string, not an alphnum string.

[DISK_SIGNATURE](https://www.yoctoproject.org/docs/2.3/ref-manual/ref-manual.html#var-DISK_SIGNATURE)

Reply:
Wrong on my end. `expr(1)` does parse a regular-expression and not a glob pattern for the second operand of `:`. My first test must have missed the `$` (`expr "x${DISK_SIGNATURE}" : "x[A-Za-z0-9]*"`) which made me think the second argument was not a regexp... My bad.